src/python/gudhi/barycenter.py

```diff
@@ -0,0 +1,224 @@
+    P = ot.emd(a=a, b=b, M=M) 
+    # Note : it seems POT returns a permutation matrix in this situation, 
+    # ie a vertex of the constraint set (generically true).
+    if withcost:
+        cost = np.sum(np.multiply(P, M))
+    P[P < 0.5] = 0  # dirty trick to avoid some numerical issues... to improve.
+    res = np.argwhere(P)
+
+    # return the list of (i,j) such that P[i,j] > 0, 
+    #i.e. x_i is matched to y_j (should it be the diag). 
+    if withcost:
+        return res, cost
+    return res
+
+
+def lagrangian_barycenter(pdiagset, init=None, verbose=False):
+    '''
+    :param pdiagset: a list of size m containing numpy.array of shape (n x 2) 
+                        (n can variate), encoding a set of 
+                        persistence diagrams with only finite coordinates. 
+                        If empty, returns None.
+    :param init: The initial value for barycenter estimate. 
+                    If None, init is made on a random diagram from the dataset. 
+                    Otherwise, it must be an int 
+                    (then we init with diagset[init])
+                    or a (n x 2) numpy.array enconding 
+                    a persistence diagram with n points.
+    :param verbose: if True, returns additional information about the
+                        barycenter.
+    :returns: If not verbose (default), a numpy.array encoding
+                    the barycenter estimate 
+                    (local minima of the energy function). 
+                    If verbose, returns a couple (Y, log)
+                    where Y is the barycenter estimate,
+                    and log is a dict that contains additional informations:
+                    - groupings, a list of list of pairs (i,j),
+                        That is, G[k] = [(i, j) ...], where (i,j) indicates 
+                        that X[i] is matched to Y[j]
+                        if i > len(X) or j > len(Y), it means they 
+                        represent the diagonal.
+                    - energy, a float representing the Frechet 
+                                energy value obtained,
+                                that is the mean of squared distances 
+                                of observations to the output.
+                    - nb_iter, integer representing the number of iterations 
+                               performed before convergence of the algorithm.
+    '''
+    X = pdiagset  # to shorten notations, not a copy
+    m = len(X)  # number of diagrams we are averaging
+    if m == 0:
+        print("Warning: computing barycenter of empty diag set. Returns None")
+        return None
+    
+    # store the number of off-diagonal point for each of the X_i
+    nb_off_diag = np.array([len(X_i) for X_i in X])  
+    # Initialisation of barycenter
+    if init is None:
+        i0 = np.random.randint(m)  # Index of first state for the barycenter
+        Y = X[i0].copy() #copy() ensure that we do not modify X[i0]
```

Comment:
Does that mean that you modify pdiagset during the algorithm? It may be worth documenting if this is the case, as the user may be confused if his diagrams change unexpectedly.

Reply:
No, sorry, maybe my line-comment is not clear. 
I do modify `Y`, so that if I wrote `Y = X[i0]`, I would also modify `X|i0]` which is not the expected behavior. Maybe it is not worth documenting on this, it is very basic.

So the answer is no : I do not modify `pdiagset` .

Reply:
Ah, yes. In retrospect, it is pretty obvious...

Reply:
Ah, yes. In retrospect, it is pretty obvious...

Reply:
I removed the line comment